feat: add Token Metrics plugin for Virtuals GAME framework - Complete…

ashuxshimra opened the pull request:
## Plugin Contribution: Token Metrics Integration

### 📊 Plugin Overview
- **Name**: Token Metrics Plugin  
- **Type**: Cryptocurrency Analysis & Trading Intelligence
- **API Integration**: 21 Token Metrics API endpoints
- **GAME Compliance**: ✅ Full compliance with Virtuals Protocol GAME framework

### 🎯 Key Features
- ✅ AI-powered cryptocurrency analysis and trading signals
- ✅ Interactive chat interface for testing and exploration  
- ✅ Comprehensive market sentiment analysis
- ✅ Quantitative metrics and technical analysis
- ✅ Real-time price data and market intelligence
- ✅ Crypto indices tracking with portfolio analysis
- ✅ 22+ example files and comprehensive test suite

### 📋 Compliance Checklist
- ✅ plugin_metadata.yml included and complete
- ✅ Comprehensive README.md with installation and usage
- ✅ Dev Testing section with screenshots and demonstrations
- ✅ Full GAME framework integration with proper error handling
- ✅ Rate limiting and robust error handling implemented
- ✅ Cross-platform compatibility verified

### 🧪 Testing & Documentation
- All 21 API endpoints tested and verified
- Interactive chat interface included for hands-on testing
- Screenshots and visual demonstrations provided
- Comprehensive documentation and developer guides included

### 📸 Visual Demonstrations
Screenshots included in `plugins/tokenMetricsPlugin/screenshots/`:
- Plugin installation and setup process
- Interactive chat interface examples  
- Market sentiment analysis output
- Quantitative metrics dashboard

This plugin provides comprehensive cryptocurrency analysis capabilities to GAME agents through the Token Metrics API integration.

ashuxshimra committed:
feat: add Token Metrics plugin for Virtuals GAME framework - Complete integration with 21 Token Metrics API endpoints, AI-powered analysis, interactive chat interface, full GAME compliance, comprehensive documentation and testing

Ang-dot commented on plugins/tokenMetricsPlugin/env.example:
```diff
@@ -0,0 +1,8 @@
+GAME_API_KEY=your-game-api-key-here
+
+# Optional: Custom TokenMetrics API Base URL (if using different endpoint)
+# TOKENMETRICS_BASE_URL=https://api.tokenmetrics.com/v2 
```
nit: add an empty line at EOF following POSIX standard

```suggestion
# TOKENMETRICS_BASE_URL=https://api.tokenmetrics.com/v2 

```

Ang-dot commented on plugins/tokenMetricsPlugin/package.json:
```diff
@@ -0,0 +1,75 @@
+    "typescript": "^5.0.0",
+    "ts-node": "^10.9.0"
+  }
+}
```
```suggestion
}

```

Ang-dot commented on plugins/tokenMetricsPlugin/plugin_metadata.yml:
```diff
@@ -0,0 +1,229 @@
+platforms:
+  - "macOS"
+  - "Linux"
+  - "Windows" 
```
```suggestion
  - "Windows" 
  
```

Ang-dot commented on plugins/tokenMetricsPlugin/src/chat.ts:
```diff
@@ -0,0 +1,4471 @@
+  } else {
+    process.exit(1);
+  }
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.aiReports.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.correlation.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.cryptoInvestors.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.dailyOhlcv.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.fullAgent.ts:
```diff
@@ -0,0 +1,35 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.hourlyOhlcv.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.hourlyTradingSignals.ts:
```diff
@@ -0,0 +1,39 @@
+  }
+}
+
+main().catch(console.error); 
```
```suggestion
main().catch(console.error); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.indices.ts:
```diff
@@ -0,0 +1,96 @@
+  console.log("=".repeat(80));
+  
+  process.exit(0);
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.indicesHoldings.ts:
```diff
@@ -0,0 +1,96 @@
+  console.log("=".repeat(80));
+  
+  process.exit(0);
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.indicesPerformance.ts:
```diff
@@ -0,0 +1,96 @@
+  console.log("=".repeat(80));
+  
+  process.exit(0);
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.investorGrades.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.marketMetrics.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.priceData.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.quantmetrics.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.resistanceSupport.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.scenarioAnalysis.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.sentiments.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.tokenmetricsAi.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})(); 
```
```suggestion
})(); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.tokens.ts:
```diff
@@ -0,0 +1,97 @@
+  console.log("=".repeat(80));
+  
+  process.exit(0);
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.topMarketCap.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.traderGrades.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/examples/example.tradingSignals.ts:
```diff
@@ -0,0 +1,39 @@
+      verbose: true,
+    });
+  }
+})();
```
```suggestion
})();

```

Ang-dot commented on plugins/tokenMetricsPlugin/src/index.ts:
```diff
@@ -0,0 +1,1101 @@
+    }
+  }
+  
+  export default TokenMetricsPlugin;
```
```suggestion
  export default TokenMetricsPlugin;
  
```

Ang-dot commented on plugins/tokenMetricsPlugin/test-all-endpoints-chat.js:
```diff
@@ -0,0 +1,257 @@
+}
+
+// Start the testing process
+runAllTests().catch(console.error); 
```
```suggestion
runAllTests().catch(console.error); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/test-routing-fixed.js:
```diff
@@ -0,0 +1,226 @@
+  console.log('\n🎉 ALL TESTS PASSED! Routing logic is working correctly for all 17 endpoints.');
+} else {
+  console.log('\n⚠️  Some tests failed. Check the routing logic above.');
+} 
```
```suggestion
} 

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-individual.ts:
```diff
@@ -0,0 +1,262 @@
+}
+
+// Run the tests
+testTokenMetricsEndpoints().catch(console.error);
```
```suggestion
testTokenMetricsEndpoints().catch(console.error);

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-integration.ts:
```diff
@@ -0,0 +1,63 @@
+  tracker.printSummary();
+}
+
+testFullIntegration().catch(console.error);
```
```suggestion
testFullIntegration().catch(console.error);

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-scenarios/indices-endpoints-demo.ts:
```diff
@@ -0,0 +1,93 @@
+  console.log(`${extendedColors.cyan}• Cross-index comparison capabilities${extendedColors.reset}\n`);
+}
+
+demoIndicesEndpoints().catch(console.error); 
```
```suggestion
demoIndicesEndpoints().catch(console.error); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-scenarios/market-analyst.ts:
```diff
@@ -0,0 +1 @@
+    
```
this test is empty, please remove if unintended or add test codes

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-scenarios/new-endpoints-demo.ts:
```diff
@@ -0,0 +1,74 @@
+  console.log(`${colors.blue}• Correlation - Token relationship analysis${colors.reset}\n`);
+}
+
+demoNewEndpoints().catch(console.error); 
```
```suggestion
demoNewEndpoints().catch(console.error); 

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-scenarios/research-agent.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+createResearchAgent().catch(console.error);
```
```suggestion
createResearchAgent().catch(console.error);

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-scenarios/trading-bot.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+createTradingBot().catch(console.error);
```
```suggestion
createTradingBot().catch(console.error);

```

Ang-dot commented on plugins/tokenMetricsPlugin/tests/test-setup.ts:
```diff
@@ -0,0 +1,140 @@
+  } else {
+    process.exit(1);
+  }
+}
```
```suggestion
}

```

Ang-dot commented on plugins/tokenMetricsPlugin/tsconfig.json:
```diff
@@ -0,0 +1,19 @@
+  },
+  "include": ["src/**/*"],
+  "exclude": ["node_modules", "dist"]
+}
```
```suggestion
}

```

ashuxshimra committed:
feat: Update TokenMetrics plugin with new APIs and POSIX compliance

✨ New Features:
- Add Moonshot Tokens API for AI-curated high-potential token picks
- Add Technology Grade API for development quality analysis
- Add TM Grade API for comprehensive grade with signals & momentum
- Add Fundamental Grade API for tokenomics and strength analysis

🗑️ Removed APIs:
- Remove getTraderGrades API (deprecated)
- Remove getInvestorGrades API (deprecated)

📚 Documentation Updates:
- Update README with comprehensive developer integration guide
- Add Quick Start section with code examples
- Add detailed Examples section for different use cases
- Update plugin metadata with new API information

🧹 Code Quality:
- Fix EOF compliance for all files (POSIX standard)
- Clean up unnecessary test files
- Update package.json scripts
- Improve error handling and logging

🔧 Technical Improvements:
- All 21 functions now available for Virtuals agents
- Natural language processing support
- Production-ready integration examples
- Comprehensive error handling

Ready for PR review! 🚀

ashuxshimra committed:
fix: update plugin metadata to comply with template requirements

Ang-dot commented on plugins/tokenMetricsPlugin/plugin_metadata.yml:
```diff
@@ -0,0 +1,19 @@
+documentation_url: "https://developers.tokenmetrics.com/"
+changelog_url: ""
+
+x_account_handle: "https://x.com/tokenmetricsinc?ref_src=twsrc%5Egoogle%7Ctwcamp%5Eserp%7Ctwgr%5Eauthor"
```
```suggestion
x_account_handle: "@tokenmetricsinc"
```

Ang-dot committed:
Update plugins/tokenMetricsPlugin/plugin_metadata.yml
